### ISSUE TITLE
Add shutdown handler for new launcher

### ISSUE THREAD

**Issue by rvl** [cardano-foundation/cardano-wallet#1314]
# Context

Part of [ADP-92](https://jira.iohk.io/browse/ADP-92) (new launcher).

## User story

As an API client (Javascript)
I want a method to cleanly and reliably stop cardano-wallet (and its node)
so that I can exit my application and release resources allocated to it.

# Decision

Implement the cardano-wallet side of the [launcher shutdown process](https://github.com/input-output-hk/cardano-launcher/blob/master/docs/windows-clean-shutdown.md).

DaedalusIPC can stay for the time being, but will be deleted later.

# Acceptance Criteria

 - The shutdown handler must work on both Windows and POSIX.

---

# Development 

- PR  #1315

# QA 

1. Unit tests in `Cardano.StartupSpec`
2. Integration test in `.../Scenario/CLI/Server.hs`.

**Comment by piotr-iohk** [comment]
hm, one of the integration tests fails for me locally. (btw. thanks for handy `PATATE` handler ;) https://github.com/input-output-hk/cardano-wallet/blob/master/lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Server.hs#L72)

```
Failures:

  test/integration/Test/Integration/Jormungandr/Scenario/CLI/Server.hs:162:17: 
  1) CLI Specifications, PATATE, LOGGING - cardano-wallet serve logging [SERIAL], LOGGING - Serve debug logs for one component
       not expected: 0

  To rerun use: --match "/CLI Specifications/PATATE/LOGGING - cardano-wallet serve logging [SERIAL]/LOGGING - Serve debug logs for one component/"

Randomized with seed 831114914

Finished in 3.8773 seconds
13 examples, 1 failure
```

**Comment by KtorZ** [comment]
Also fails for me locally for a reason I can't explain, but does not on CI :thinking:

**Comment by piotr-iohk** [comment]
@rvl, @KtorZ:
 - is this failure on local machine (https://github.com/input-output-hk/cardano-wallet/issues/1314#issuecomment-594061933) is something that should be investigated further?

 - also acceptance criteria for ADP-92 is quite extensive:
```
1. I want to start cardano-wallet (and its node)
2. I want to cardano-wallet to use a free TCP port for its server
3. I want to receive events when the wallet backend is started/stopped
4. I want to receive events when the node is started/stopped
5. I want to receive an event when the wallet API is ready
6. I want a method to get the wallet backend API base URL and connection parameters
7. I want a method to cleanly and reliably stop cardano-wallet (and its node)
8. I want a documented Javascript module with typed interface files
9. I want the cardano-launcher to be well-tested on Windows
10. I want the module to output detailed logging
```
Does this ticket cover all points or part (which part if so)?
 - Also in particular re pt. 9 from AC - do you recommend any manual tests to perform under Windows (or any further manual or automated tests) to be performed or added

**Comment by rvl** [comment]
Sorry @piotr-iohk I lost track of your message.

Yes the failure is due to `withShutdownHandler` and I am working on fixing it.

Those items from ADP-92 are user stories and each has its own AC. It's a bit difficult to track because the user stories are implemented in two git repos - cardano-launcher and cardano-wallet.

This ticket covers "7. I want a method to cleanly and reliably stop cardano-wallet (and its node)"

Regarding point 9 from the user stories - there are some automated tests for windows in the cardano-launcher. The end-to-end test is still pending on input-output-hk/cardano-launcher#12.

**Comment by piotr-iohk** [comment]
Ok, the failing test is fixed but there is now also this opened PR related to this task #1424. (Not sure than, shall we move it back to "in progress" or...?)

**Comment by rvl** [comment]
I don't know ... the code removal is not part of the epic, it's just a clean up which I wanted to do early because its tests are fragile. I'm not sure when to make a github issue and when not, because we have jira there, and also multiple github repos, ...

**Comment by piotr-iohk** [comment]
All right. It seems the original scope for this particular GH task is addressed, so I will close this one.
Then maybe if there are further parts related to ADP-92 that need QA review, let's add them as separate GH tasks moving forward.
